contracts/SurplusBeneficiary.sol

```diff
@@ -0,0 +1,131 @@
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import { AddressUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+import { SafeMathUpgradeable } from "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import {
+    SafeERC20Upgradeable,
+    IERC20Upgradeable
+} from "@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { SafeMathUpgradeable } from "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import { FullMath } from "@uniswap/v3-core/contracts/libraries/FullMath.sol";
+import { SurplusBeneficiaryStorageV1 } from "./storage/SurplusBeneficiaryStorage.sol";
+import { ISurplusBeneficiary } from "./interface/ISurplusBeneficiary.sol";
+import { IFeeDistributor } from "./interface/IFeeDistributor.sol";
+
+contract SurplusBeneficiary is
+    ISurplusBeneficiary,
+    ReentrancyGuardUpgradeable,
+    OwnableUpgradeable,
+    SurplusBeneficiaryStorageV1
+{
+    using AddressUpgradeable for address;
+    using SafeMathUpgradeable for uint256;
+
+    /// @dev this function is public for testing
+    // solhint-disable-next-line func-order
+    function initialize(
+        address tokenArg,
+        address feeDistributorArg,
+        address treasuryArg,
+        uint24 treasuryPercentageArg
+    ) public initializer {
+        __ReentrancyGuard_init();
+        __Ownable_init();
+
+        setToken(tokenArg);
+        setFeeDistributor(feeDistributorArg);
+        setTreasury(treasuryArg);
+        setTreasuryPercentage(treasuryPercentageArg);
+
+        IERC20Upgradeable(_token).approve(feeDistributorArg, uint256(-1));
+    }
+
+    //
+    // PUBLIC ONLY-ADMIN NON-VIEW
+    //
+
+    function setToken(address tokenArg) public onlyOwner {
+        // SB_TANC: token address is not contract
+        require(tokenArg.isContract(), "SB_TANC");
+        emit TokenChanged(_token, tokenArg);
+        _token = tokenArg;
+    }
+
+    function setFeeDistributor(address feeDistributorArg) public onlyOwner {
+        // SB_FDNC: feeDistributor address is not contract
+        require(feeDistributorArg.isContract(), "SB_FDNC");
+        emit FeeDistributorChanged(_feeDistributor, feeDistributorArg);
+        _feeDistributor = feeDistributorArg;
+    }
+
+    function setTreasury(address treasuryArg) public onlyOwner {
+        // SB_TNC: treasury address is not contract
+        require(treasuryArg.isContract(), "SB_TNC");
+        emit TreasuryChanged(_treasury, treasuryArg);
+        _treasury = treasuryArg;
+    }
+
+    function setTreasuryPercentage(uint24 treasuryPercentageArg) public onlyOwner {
+        // SB_TPZ: treasury percentage is equal to zero
+        require(treasuryPercentageArg != 0, "SB_TPZ");
```

Comment:
why can't be 0?
should check it's <= 1e6

Reply:
fixed